core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -774,10 +779,15 @@ public static Expression translateLiteral(
         return Expressions.constant(bd, javaClass);
       }
       assert javaClass == BigDecimal.class;
-      return Expressions.new_(BigDecimal.class,
-          Expressions.constant(
+      Expression expression =
+          Expressions.new_(
+              BigDecimal.class, constant(
               requireNonNull(bd,
                   () -> "value for " + literal).toString()));
+      if (type.getScale() != bd.scale()) {
+        expression = adjustDecimalScale(type.getScale(), expression);
+      }
+      return expression;
```

Comment:
If we have a literal why do we need to adjust the scale? Does the `BigDecimal(String)` constructor pick a wrong scale for certain inputs? If that's the case can you please provide some examples with such inputs?

Which of the existing unit tests are affected by this change?

Reply:
when invoke RexBuilder#canRemoveCastFromLiteral the call, for example `cast(-0.1 as decimal(19, 0))` returns true, then it will be translate to -0.1:BigDeciaml(19, 0) as a literal. 
![image](https://github.com/apache/calcite/assets/13960353/127547ab-b0d1-4389-9b7c-17084340be9d)
However, the BigDecimal gives scale=1. So we should change the scale to the correct one.
If we run the `CalciteSqlOperatorTest.java` we will found this problem, but the agg.iq, misc.iq, and the expected values of `SqlOperatorTest.java` are wrong, so this issue is being ignored.

Reply:
The fact that we have a literal that is `-0.1:DECIMAL(19,0)` seems like a bug to me that we may need to address first.

These kind of literals indicate that there is a problem at the logical layer but this PR attacks the problem at the physical layer which is not very intuitive.

Moreover, we want to fix the root cause so that users who are not using at all Calcite's physical layer (Enumerable, Bindable, code generation, etc.) can benefit from the fix.

Reply:
I think we can keep the original proposal and deal with `-0.1:DECIMAL(19,0)` in a follow-up.